README.md
# OKOVISION V2

Interface web de supervison d'une chaudiere Okofen avec firmware V4.00b

<http://okovision.dronek.com> (pour la V1)


changelog.md
Unrealised
----------

2.00.b
------
* Make Okovision compatible with firmware V4.

1.10.0
------
* Make Okovision International by adding English language (default).
  * For changing Language, go to parameters menu
* This is the last update for the coming months!  

1.9.2
------
* Fix #108 - Error Message rendu.getIndicByDay when you don' have pumpe Hot water


1.9.1
------
* Fix #107 - Error Message after updating. In this update you still have the error. But not next update
* Fix script install with new data Hot Water

1.9.0
-----

* Add #85 : 
  * Display of domestic hot water consumption for the current day, and into summary reports
  * To more data on the previous days, you must go to the menu "Calculation of daily summaries / Calcul des synthèse jouralières" and force recalculation

1.8.5
-----

* Fix #37 : Improved Cron function - Get ALL CSV on boiler if not yet imported

1.8.4
-----

* Okovision V2 Foundation : Cleanning Code
* Fix #95 : Import from CSV file Firmware V3 doesn't work (thank's John47 !)
* Improvement Wiki Documentation

1.8.3
-----

* Refactoring du code pour respecter l'indentation PHP
* Merge Fix84 proposé par bertrandgorge : Include all files with absolute paths
* Merge Fix82 proposé par grouxx : Gestion dans GetIndic de plusieurs circuite + ajout de nouveaux retours
* Merge Fix81 proposé par grouxx : Correction get and set boiler Mode pour choisir le bon circuit

1.8.2
-----
* Version compatible PHP 5.6 et PHP 7.2
* Correction de compatibilité avec Mysql 5.7 et MariaDB 10.3

1.8.1
-----

* Correction anomalie lors du setup pour le choix IP / USB
* Correction déconnexion sous FireFox

1.8.0
-----

* Correction orthographe
* Ajout gestion du stock des pellets (Silo et sac)
* Gestion des evenements de la chaudière (ramonage, vidage du cendrier, entretien)

1.7.4
-----

* Correction perte des noms des capteurs suite maj 1.7.3

1.7.3
-----

* Correction mineur dans initialisation de la base de donnée (bgorge)
* Quand un utilisateur import pour la meme journée le csv en http et CSV, les données ne sont plus en double (bgorge)

1.7.2
-----

* Ajout d'une API permettant à des applications tiers d'interagir avec okovision et la chaudière

1.7.1
-----

* Ajout d'une alerte pour dire de sauvegarder si changement de parametres
* Prise en compte d'installation multi-chaudiére (le calcul de la conso ne sera que pour la chaudière maitre)
* Revision de la methode d'installation -> optimisation de l'espace de stockage necessaire pour les données journalières
* Graphe synthèse saison, ne pas voir le mois en cours

1.7.0
-----

* Correction orthographe / correction synthaxique
* Changement de l'unité ms en ds pour la vis sans fin
* Calendrier lors de la selection de la date sur la page d'accueil
* Afficher les noms des courbes dans l'ordre des courbes
* Ajout page "Temps Réel"
    *  Visualisation des parametres de réglage combustion chaudière et régulation
    *  Visualisation des graphiques en temps réel
    *  Sauvegarde de la configuration de la chaudiere
    *  Modification des parametres de la chaudiere via okovision
    *  Rechargement de parametres sauvegardées et modification des parametres de la chaudière
    *  Visualisation sur les graphes journaliers de la modification des parametres de la chaudière
*  Suppression de la matrice sans perte de données de l'historique (mais suppresion des données journalières)

1.6.4
-----

* préparation livraison 1.7.0
* correction orthographe / correction synthaxique

1.6.2
-----

* Correction setup (compte admin non present)
* Ajout colonne DJU dans tableau recap
* Maj du texte dans 'A propos'

1.6.1
-----

* Correction upload fichier en erreur
* Correction setup impossible

1.6.0
-----

* Creation d'un espace membre contenant la configuration de l'application (defaut admin/okouser)
* Ajout de page d'erreur
* Creation d'un .htaccess

1.5.5
-----

* Ajout d'une alert growl en page d'index pour un maj disponible

1.5.4
-----

* Correction definitive du probleme de fuseau horaire
* Correction probleme d'encodage lors de la creation de la matrice sur linux
* Mise en place Y axe min dynamique (par defaut 0 ou alors valeur negative)


1.5.3
-----
* ajout d'un parametre dans hightchart pour ne pas appliquer un offset sur le timestamp en fonction du navigateur. Force l'utilisation d'UTC

1.5.0
-----
* Possibilité de recalculer la synthese sur une periode choisie
* Mise à jour de la matrice possible sans perte de données
* Petites retouches ergonomiques
* Refraichir le numero de version après un l'installation d'une maj

1.4.3
-----

* Ajout du choix du fuseau Horaire

1.4.0
-----

* Tableau gr/dju/m2
* Sync zoom graphe + maj indicateur haut de page sur la zone séléctionnée
* Optimisation rendu graphe journalier
* Utilisation du status 4 et maj de la bdd

1.3.0
-----

* #26 - Refonte du modele de la base pour réduire son volume
* #26 - Creation d'un script de migration des données (lien disponible dans la page 'A propos')
* #27,#28,#29 - Réécriture du code impacté par le changement du modele de données
* #30 - correction anomalie sur calcul synthese lancé via Cron
* Optimisation des performances + gestion pool de connexion bdd
* Redécoupage des pages d'administration

1.2.1
-----

* #25 - voir la version courante
* #19 - anomalie html sur la page des historiques
* Amelioration du calcul global journalier
* #21 - import de masse

1.1.1
-----

* #23 - synthse journalière ne fonctionnait plus

1.1.0
------

* #14 et #11 - EVOL - Externalisation des textes dans un fichier commun (pour gain de perf et internationnalisation si besoin)
* #11 - EVOL -  Factorisation des appels Asynchrones - Uniformisation / performance / évolutions futur facilitée
* #2 et #18 - EVOL - Gestion de la position des graphes et de capteurs dans les graphes
* #12 - FIX - Page "A propos" - bouton encore visible après l'installation de la mise à jour
* #6 - FIX - Liste deroulante des saisons preselectionnées sur la periode en cours

setup.php
<?php

    /**
     * Projet : Okovision - Supervision chaudiere OeKofen
     * Auteurs : Stawen Dronek mod by skydarc for V2
     * Utilisation commerciale interdite sans mon accord.
     */
    function is_ajax()
    {
        return isset($_SERVER['HTTP_X_REQUESTED_WITH']) && 'xmlhttprequest' == strtolower($_SERVER['HTTP_X_REQUESTED_WITH']);
    }

    function testBddConnection($s)
    {
        mysqli_report(MYSQLI_REPORT_STRICT);

        $r = true;

        try {
            $db = new mysqli($s['db_adress'], $s['db_user'], $s['db_password']);
        } catch (Exception $e) {
            $r = false;
        }
        $t['response'] = $r;
        header('Content-type: text/json');
        echo json_encode($t, JSON_NUMERIC_CHECK);

        exit(23);
    }

    function makeInstallation($s)
    {
        if ($s['createDb']) {
            // create BDD
            $mysqli = new mysqli($s['db_adress'], $s['db_user'], $s['db_password']);

            // check connection
            if ($mysqli->connect_errno) {
                printf("Connect failed: %s\n", $mysqli->connect_error);
                exit(24);
            }

            $q = 'CREATE DATABASE IF NOT EXISTS `'.$s['db_schema'].'` /*!40100 DEFAULT CHARACTER SET utf8 */;';
            if (!$mysqli->query($q)) {
                echo 'Création BDD impossible';
                exit;
            }
            $mysqli->close();
        }

        $mysqli = new mysqli($s['db_adress'], $s['db_user'], $s['db_password'], $s['db_schema']);

        // execute multi query
        $mysqli->multi_query(file_get_contents('install/install.sql'));
        while ($mysqli->next_result()) {
        } // flush multi_queries

        // init de la table des dates de reference
        $start_day = mktime(0, 0, 0, 9, 1, 2014); //1er septembre 2014
        $stop_day = mktime(0, 0, 0, 9, 1, 2037); //justqu'au 1er septembre 2037, on verra en 2037 si j'utilise encore l'app.
        $nb_day = ($stop_day - $start_day) / 86400;
        $query = 'INSERT INTO oko_dateref (jour) VALUES ';
        for ($i = 0; $i <= $nb_day; ++$i) {
            $day = date('Y-m-d', mktime(0, 0, 0, date('m', $start_day), date('d', $start_day) + $i, date('Y', $start_day)));
            $query .= "('".$day."'),";
        }

        $query = substr($query, 0, strlen($query) - 1).';';

        $mysqli->query($query);

        $mysqli->close();

        // Make Config.php
        $configFile = file_get_contents('config_sample.php');

        $configFile = str_replace('###_BDD_IP_###', $s['db_adress'], $configFile);
        $configFile = str_replace('###_BDD_USER_###', $s['db_user'], $configFile);
        $configFile = str_replace('###_BDD_PASS_###', $s['db_password'], $configFile);
        $configFile = str_replace('###_BDD_SCHEMA_###', $s['db_schema'], $configFile);

        $configFile = str_replace('###_CONTEXT_###', getcwd(), $configFile);

        $configFile = str_replace('###_TOKEN_###', sha1(rand()), $configFile);
        //$configFile = str_replace("###_TOKEN-API_###",sha1(rand()),$configFile);

        file_put_contents('config.php', $configFile);

        // Make config.json
        $param = [
            'chaudiere' => $s['oko_ip'],
            'port_json' => $s['oko_json_port'],
			'password_json' => $s['oko_json_pwd'],
			'url_mail' =>  $s['mail_host'],
			'login_mail' =>  $s['mail_log'],
			'password_mail' =>  $s['mail_pwd'],
            'tc_ref' => $s['param_tcref'],
            'poids_pellet' => $s['param_poids_pellet'],
            'surface_maison' => $s['surface_maison'],
            'get_data_from_chaudiere' => $s['oko_typeconnect'],
            'send_to_web' => '0',
            'has_silo' => '0',
            'lang' => 'en',
        ];

        file_put_contents('config.json', json_encode($param));
    }

    if (is_ajax()) {
        if (isset($_GET['type'])) {
            switch ($_GET['type']) {
                case 'connect':
                    testBddConnection($_POST);

                    break;
                case 'install':
                    makeInstallation($_POST);

                    break;
            }
        }
    }

?>

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>OkoVision</title>
    
	<!-- Bootstrap core CSS -->
    <link href="css/bootstrap.min.css" rel="stylesheet">
    <!-- Bootstrap theme -->
    <link href="css/bootstrap-theme.min.css" rel="stylesheet">
    <link href="css/jquery-ui.min.css" rel="stylesheet">
    <link href="css/custom.css" rel="stylesheet">
    <link href="css/animate.css" rel="stylesheet">
    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>+
    <![endif]-->
	<?php //include_once("analyticstracking.php");?>
	
	</head>

  <body role="document">
  	
<div class="container theme-showcase" role="main">
		<div class="page-header" align="center">
			<h2>Okovision installation</h2> <br>
		</div>
		<div>
			<h3><small>You can modify this information, after installation, through the settings screen</small></h3>
		</div>
		
		
			<fieldset>
				<form class="form-horizontal" id="formConnect">
				<!-- Form Name -->
					<legend>Database Connection</legend>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="db_adress">Address (*) :</label>  
					  <div class="col-md-3">
					  <input id="db_adress" name="db_adress" type="text" placeholder="ex : localhost, 192.168.xxx.xxx" class="form-control input-md" required="">
					  <span class="help-block"></span>  
					  </div>
					</div>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="db_adress">Name (*) :</label>  
					  <div class="col-md-3">
					  <input id="db_schema" name="db_schema" type="text" placeholder="ex : okovision" class="form-control input-md" required="">
					  <span class="help-block"></span>  
					  </div>
					</div>
					<div class="form-group">
						<label class="col-md-4 control-label" for="createDb">Create database :</label>  
					  	<div class="col-md-3 checkbox">
					    <label>
					      <input id="createDb" type="checkbox"> don't checked if database already exist
					    </label>
					
					  </div>
					</div>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="db_user">User (*) :</label>  
					  <div class="col-md-3">
					  <input id="db_user" name="db_user" type="text" placeholder="ex: root" class="form-control input-md" required="">
					  </div>
					</div>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="db_password">Password (*) :</label>  
					  <div class="col-md-3">
					  <input id="db_password" name="db_password" type="text" placeholder="ex : toor" class="form-control input-md" required="">
					  </div>
					</div>
					
					<!-- Button -->
					<label class="col-md-4 control-label"  for="bt_testConnection">Connection test :</label>
					  <div class="col-md-3">
					    <button id="bt_testConnection" name="bt_testConnection" class="btn btn-primary" type="button">Test</button>
					  </div>
					</form>
			</fieldset>
			

			<form class="form-horizontal">
				<fieldset>
				
				<!-- Form Name -->
					<legend>Boiler Communication</legend>
					
					<!-- Select Basic -->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="oko_typeconnect">CSV file grab mode :</label>
					  <div class="col-md-3">
					    <select id="oko_typeconnect" name="oko_typeconnect" class="form-control">
					        <option value="0">USB</option>
			                <option value="1">IP</option>
							<option value="2">IP via Json (firmware v4.00b)</option>
					    </select>
					  </div>
					</div>
					
					<!-- Text input-->
					<div class="form-group" id="form-ip" style="display: none;">
					  <label class="col-md-4 control-label" for="oko_ip">Boiler IP address :</label>  
					  <div class="col-md-3">
					    <input id="oko_ip" name="oko_ip" type="text" placeholder="ex : 192.168.0.xx" class="form-control input-md">
					  </div>
					</div>
					
					<div class="form-group" id="form-json-port" style="display: none;">
					  <label class="col-md-4 control-label" for="oko_json_port">Json port :</label>  
					  <div class="col-md-3">
					    <input id="oko_json_port" name="oko_json_port" type="text" placeholder="ex : 4321" class="form-control input-md">
					  </div>
					</div>
					
					<div class="form-group" id="form-json-pwd" style="display: none;">
					  <label class="col-md-4 control-label" for="oko_json_pwd">Json password :</label>  
					  <div class="col-md-3">
					    <input id="oko_json_pwd" name="oko_json_pwd" type="text" placeholder="ex : A1b2" class="form-control input-md">
					  </div>
					</div>
				
				</fieldset>
				
					<div id="form-mail" style="display: none;">
						<legend>Recovery of csv from mailbox</legend>
    					
							<div class="form-group" id="form-host">
								<label class="col-md-4 control-label" for="oko_ip">MailBox server host (<a href='https://www.php.net/manual/en/function.imap-delete.php' style='text-decoration; none; color; inherit;'>info</a>)</label>  
								<div class="col-md-3">
									<input id="mail_host" name="mail_host" type="text" class="form-control input-md" placeholder="ex : 127.0.0.1">
								</div>
							</div>
						
							<div class="form-group" id="form-login">
								<label class="col-md-4 control-label" for="oko_ip">Mailbox login</label>  
								<div class="col-md-3">
									<input id="mail_log" name="mail_log" type="text" class="form-control input-md" placeholder="login">
								</div>
							</div>
							
							<div class="form-group" id="form-loginPwd">
								<label class="col-md-4 control-label" for="oko_ip">Mailbox password</label>  
								<div class="col-md-3">
									<input id="mail_pwd" name="mail_pwd" type="text" class="form-control input-md" placeholder="password">
								</div>
							</div>
					</div>
    				
    				
			</form>

			<form class="form-horizontal">
				<fieldset>
				
				<!-- Form Name -->
					<legend>Application settings</legend>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="param_tcref">Reference °C :</label>  
					  <div class="col-md-3">
					  <input id="param_tcref" name="param_tcref" type="text" placeholder="ex : 20" class="form-control input-md" required="" value="20">
					  <span class="help-block">If you have 2 setpoints, reduced to 19°C and comfort at 21°C, you average -&gt; 20°C. It's for DJU calculation</span>  
					  </div>
					</div>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="param_poids_pellet">Pellet weight for 60 seconds of work : </label>  
					  <div class="col-md-3">
					  <input id="param_poids_pellet" name="param_poids_pellet" type="text" placeholder="ex : 150" class="form-control input-md" required=""  value="150">
					  <span class="help-block">Pellet weight in grams measured by operating the furnace feed screw for 60 seconds</span>  
					  </div>
					</div>
					
					<!-- Text input-->
					<div class="form-group">
					  <label class="col-md-4 control-label" for="parap_poids_pellet">House surface : </label>  
					  <div class="col-md-3">
					  <input id="surface_maison" name="param_surface" type="text" placeholder="ex : 180" class="form-control input-md" required=""  value="180">
					  <span class="help-block">in m²</span>  
					  </div>
					</div>
				
				</fieldset>
			</form>
            
            	<!-- Button -->
					
					  <div class="col-md-12" align="center">
					    <button id="bt_install" name="bt_install" class="btn btn-primary" type="button">Install</button>
					  </div>

	 </div> <!-- /container -->
	
	 <!-- Bootstrap core JavaScript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
  <script src="js/jquery/jquery.min.js"></script>
	<script src="js/jquery/jquery-ui.min.js"></script>
	<script src="js/bootstrap/bootstrap.min.js"></script>
	<script src="js/bootstrap/bootstrap-notify.min.js"></script>
	<script src="js/highstock/highstock.js"></script>
	
	<script src="_langs/fr.text.js"></script>
	<script src="js/custom.js"></script>
<!--appel des scripts personnels de la page -->
	<script src="js/setup.js"></script>
    </body>
</html>